internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -71,98 +89,73 @@ func walkFn(source, dest string, recursive bool, reader afero.Fs, upload UploadF
 		if !ok {
 			return nil
 		}
-		file, err := reader.Open(path)
+
+		hash := sha256.New()
+		buf := &bytes.Buffer{}
+		file, err := u.FS.Open(path)
 		if err != nil {
-			return fmt.Errorf("open file on path %q: %w", path, err)
+			return fmt.Errorf("open %q: %w", path, err)
 		}
 		defer file.Close()
-		fileRel, err := filepath.Rel(source, path)
+
+		_, err = io.Copy(io.MultiWriter(buf, hash), file)
 		if err != nil {
-			return fmt.Errorf("get relative path for %q against %q: %w", path, source, err)
+			return fmt.Errorf("copy %q: %w", path, err)
 		}
-		key := filepath.Join(dest, fileRel)
-		url, err := upload(key, file)
+
+		// rel is "." when sourcePath == path
+		rel, err := filepath.Rel(sourcePath, path)
 		if err != nil {
-			return fmt.Errorf("upload file %q to destination %q: %w", path, key, err)
+			return fmt.Errorf("get relative path for %q against %q: %w", path, sourcePath, err)
 		}
-		*urlsPtr = append(*urlsPtr, url)
-		return nil
-	}
-}
 
-type filepathMatcher interface {
-	match(path string) (bool, error)
-}
-
-type reincludeMatcher string
+		dest := filepath.Join(destPath, rel)
+		if dest == "." { // happens when sourcePath is a file and destPath is unset
+			dest = info.Name()
+		}
 
-func buildReincludeMatchers(reincludes []string) []filepathMatcher {
-	var matchers []filepathMatcher
-	for _, reinclude := range reincludes {
-		matchers = append(matchers, reincludeMatcher(reinclude))
+		*assets = append(*assets, asset{
+			localPath:          path,
+			content:            buf,
+			ArtifactBucketPath: filepath.Join(u.PathPrefix, hex.EncodeToString(hash.Sum(nil))),
+			ServiceBucketPath:  dest,
+		})
+		return nil
 	}
-	return matchers
-}
-
-func (m reincludeMatcher) match(path string) (bool, error) {
-	return match(string(m), path)
 }
 
-type excludeMatcher string
+func (u *ArtifactBucketUploader) uploadAssets(assets []asset) error {
+	g, _ := errgroup.WithContext(context.Background())
 
-func buildExcludeMatchers(excludes []string) []filepathMatcher {
-	var matchers []filepathMatcher
-	for _, exclude := range excludes {
-		matchers = append(matchers, excludeMatcher(exclude))
+	for i := range assets {
+		asset := assets[i]
+		g.Go(func() error {
+			if err := u.Upload(asset.ArtifactBucketPath, asset.content); err != nil {
```

Comment:
problem for another day: I wonder how we can show progress bars while uploading these assets to s3